feat(matrix): add in-place matrix transposition for non-square matrix

eholum opened the pull request:
Close https://github.com/evenfurther/pathfinding/issues/497.

I took a stab at the most basic implementation described in https://en.wikipedia.org/wiki/In-place_matrix_transposition. There are definitely more optimal approaches given the references from that page, as this requires MN additional memory for recording visited locations.

Given the added complexity and memory, would it be preferable to fork the logic here for square vs non-square matrices?

samueltardieu commented:
Thanks for your contribution, it looks very interesting.

> Given the added complexity and memory, would it be preferable to fork the logic here for square vs non-square matrices?

It would be interesting to add a benchmark specifically for transposing a large enough square matrix, and see how switching to this algorithm introduces a performance regression or doesn't (in which case it would not be needed to specialize it for square matrices). Also, it would be interesting to see whether using the [bitvec](https://crates.io/crates/bitvec) makes performance better.

Would you be willing to do that? Otherwise I'll use your patch as a starting base (keeping the attribution) and work on this on the next weeks.

eholum commented:
Yes, I can took a look at that! I'll start by adding a benchmark and comparing.

If I have the time I might take a look at using [this approach from Catanzaro et. al.](https://dl.acm.org/doi/10.1145/2692916.2555253), and seeing how that compares, too.

eholum commented:
Will leave this as a draft until https://github.com/evenfurther/pathfinding/pull/526 is complete and I generate some data for bench marking.

eholum commented:
Running the matrix benchmark locally is pretty telling:

```

eholum commented:
`

eholum commented:
Updates:

* Rebased over the latest main and fixed commit messages
* Given the 8x slowdown for square matrices, I forked the in place transpose logic for square vs non square matrices
  * The benchmark time above is accurate for square matrix transposition using `transpose_in_place_square` vs `transpose_in_place_non_square`
  * We could probably make this _slightly_ faster if we required `C: Default`, since we could store temporary values and avoid some of the in place values - but I think this breaks the notion of "in place".
* Added a bench for non square matrix transposition
* Switch to using `bitvec` for tracking visited places more efficiently
  * `bitvec` adds latency to the benchmark (`58.985 µs` vs `45.935 µs `). I think this makes sense. The bit operations are more memory efficient, but adds computational complexity with bitwise operations. Given that this is supposed to be an "in place" swap, maybe it makes sense to consume as little extra memory as possible at the expense of increased computational complexity.
 
Overall my main learning is that in-place transposition of a non-square matrix is a complex operation! If I have the time I will implement the Catanzaro approach to see how it compares. If speed is the only concern and you have plenty of memory, then  using `transposed` is much faster.

samueltardieu commented:
Thanks, this looks fine!

eholum commented:
Lint checkers changing underneath us! I believe https://github.com/evenfurther/pathfinding/pull/527 will resolve the clippy failure from the merge queue run if you would like to resolve that separately?

samueltardieu commented:
This has been released as part of pathfinding 4.9.0. Note that I replaced bitvec by a simple bit manipulation expression, as this is more efficient and does not require the `wyz` and `tap` phantom dependencies.

Thanks again.
